src/devices/Board/Board.cs

```diff
@@ -0,0 +1,692 @@
+        /// <param name="defaultNumberingScheme">Default numbering scheme for the board.</param>
+        /// <remarks>
+        /// The constructor will never throw an exception. Call <see cref="Initialize"/> to initialize the hardware and check
+        /// whether an instance can actually run on the current hardware.
+        /// </remarks>
+        protected Board(PinNumberingScheme defaultNumberingScheme)
+        {
+            _defaultNumberingScheme = defaultNumberingScheme;
+            _pinReservations = new Dictionary<int, List<PinReservation>>();
+            _i2cBuses = new Dictionary<int, I2cBusManager>();
+            _pinReservationsLock = new object();
+            _initialized = false;
+            _disposed = false;
+        }
+
+        /// <summary>
+        /// True if the board instance is initialized
+        /// </summary>
+        protected bool Initialized => _initialized;
+
+        /// <summary>
+        /// True if this instance is disposed.
+        /// Any attempt to use it after this becomes true results in undefined behavior.
+        /// </summary>
+        protected bool Disposed
+        {
+            get
+            {
+                return _disposed;
+            }
+        }
+
+        /// <summary>
+        /// The default pin numbering scheme
+        /// </summary>
+        public PinNumberingScheme DefaultPinNumberingScheme
+        {
+            get
+            {
+                return _defaultNumberingScheme;
+            }
+        }
+
+        /// <summary>
+        /// Converts pin numbers from one numbering scheme to another.
+        /// </summary>
+        /// <param name="pinNumber">Pin number to convert</param>
+        /// <param name="inputScheme">The numbering scheme of the input pin</param>
+        /// <param name="outputScheme">The desired numbering scheme</param>
+        /// <returns>The converted pin number</returns>
+        public abstract int ConvertPinNumber(int pinNumber, PinNumberingScheme inputScheme, PinNumberingScheme outputScheme);
+
+        /// <summary>
+        /// Reserves a pin for a specific usage. This is done automatically if a known interface (i.e. GpioController) is
+        /// used to open the pin, but may be used to block a pin explicitly, i.e. for UART.
+        /// </summary>
+        /// <param name="pinNumber">The pin number, in the boards default numbering scheme</param>
+        /// <param name="usage">Intended usage of the pin</param>
+        /// <param name="owner">Class that owns the pin (use "this")</param>
+        /// <exception cref="InvalidOperationException">The pin is already reserved</exception>
+        public virtual void ReservePin(int pinNumber, PinUsage usage, object owner)
+        {
+            if (!_initialized)
+            {
+                Initialize();
+            }
+
+            int logicalPin = ConvertPinNumber(pinNumber, DefaultPinNumberingScheme, PinNumberingScheme.Logical);
+
+            lock (_pinReservationsLock)
```

Comment:
Is it possible to avoid this lock and instead just use ConcurrentDictionary for the pinReservations?

Reply:
The logic is not straight-forward, and it's only used when opening/closing pins, therefore performance shouldn't be an issue.